### PR TITLE
Compose StructuredTraceGraph from Entities and events graph

### PR DESCRIPTION
**Pull request by rish691**
Now that we are caching the graph in the enricher, this change gives finer control on what part of the graph to rebuild on changes in trace

### CODE REVIEW

**Comment by kotharironak** [comment]
will help in case 4 of issue - https://github.com/hypertrace/hypertrace/issues/244

**Review by skjindal93 — DISMISSED** [review]
Looks good. 

There are certain important lines which aren't covered by the unit tests. Can you please look at the codecov warnings and add unit tests for the same, if possible?

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java [hunk 1]
```diff
@@ -20,187 +8,41 @@
...
+    return graph;
   }
 
-  /**
-   * @return an immutable map of event ids to events
-   */
-  public Map<ByteBuffer, Event> getEventMap() {
-    return eventMap;
-  }
+  public static StructuredTraceGraph reCreateTraceEntitiesGraph(StructuredTrace trace) {
```
Comment (kotharironak) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:32]: How are we planning to use this downstream in ingester with thread-local - StructuredTraceGraphBuilder?
So, we don't want to build an event graph every time the structure trace has changed due to a newly added/updated entity?

Comment (kotharironak) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:32]: So, I think, we will check if trace has modified for entity or event, and accordingly reCreate the state, right?

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import org.hypertrace.core.datamodel.StructuredTrace;
+
+public class TraceEntitiesGraph {
+
+  /* entity have many-to-many relationship */
+  private final Map<String, List<Entity>> parentToChildrenEntities;
+  private final Map<String, List<Entity>> childToParentEntities;
+
+  /* these containers should be unmodifiable after initialization as we're exposing them via getters */
+  private Map<ByteBuffer, Event> eventMap;
```
Comment (kotharironak) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java:23]: This  `eventMap` is unused, and I think it is not required.

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java [hunk 3]
```diff
@@ -20,187 +8,41 @@
...
-   * @return an immutable set containing the root events
-   */
-  public Set<Event> getRootEvents() {
-    return rootEvents;
-  }
-
-  /**
-   * @return an immutable set containing the root entities
-   */
-  public Set<Entity> getRootEntities() {
```
Comment (kotharironak) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:67]: As we are changing the interface, we will need to change the consumer, do we need StructureTraceGraph ? Or should we just keep the interface as is, and just fetch the data either from internal TraceEventGraph or TraceEntityGraph?

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java (hunk 1 above)
Comment (rish691) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:32]: yes we check which one of Event/Entities graph needs rebuilding

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java (hunk 3 above)
Comment (rish691) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:67]: What challenges do you see with changing the consumer?

Comment (kotharironak) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:67]: Now, `StructureTraceGraph` looks dummy holder for `TraceEventGraph` and `TraceEntityGraph`. So, I think, either we can remove it completely or we can maintain its existing public interface as-is for downstream consumption. So, all its public methods will internally be served via either `TraceEventGraph` or `TraceEntityGraph`. And both of them will become private in the context of `StructureTraceGraph`. And, we will only change in downstream `StructureTraceGraphBuiler` which will only build portion - either `TraceEventGraph` or `TraceEntityGraph` - on trace change.

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    this.parentToChildrenEvents = new HashMap<>();
+  }
+
+  /**
+   * Create the instance with or without the full traversal graph includesFullGraph    Includes
+   * computing the full traversal graph or not
+   */
+  static TraceEventsGraph createGraph(StructuredTrace trace) {
+    TraceEventsGraph graph = new TraceEventsGraph();
+    graph.buildEventMap(trace);
```
Comment (avinashkolluru) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java:38]: Both buildEventMap and buildRootEvents seem to be traversing through all the events.
We can save some time by moving calling buildParentChildRelationship first and then do both buildEventMap and buildRootEvents in one pass

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    this.parentToChildrenEntities = new HashMap<>();
+  }
+
+  /**
+   * Create the instance with or without the full traversal graph includesFullGraph    Includes
+   * computing the full traversal graph or not
+   */
+  public static TraceEntitiesGraph createGraph(StructuredTrace trace) {
+    TraceEntitiesGraph graph = new TraceEntitiesGraph();
+    graph.buildEntityMap(trace);
```
Comment (avinashkolluru) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java:35]: I think my comment on TraceEventsGraph holds true here too

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java [hunk 6]
```diff
@@ -20,187 +14,81 @@
...
+    graph.traceEntitiesGraph = TraceEntitiesGraph.createGraph(trace);
+
+    return graph;
+  }
+
+  public static StructuredTraceGraph reCreateTraceEventsGraph(StructuredTrace trace) {
+    if (null == graph) {
+      return createGraph(trace);
+    }
+    graph.traceEventsGraph = TraceEventsGraph.createGraph(trace);
```
Comment (avinashkolluru) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java:34]: Case where graph is null, we seem to be calling createGraph which internally does `TraceEventsGraph.createGraph(trace)`. So this seems like a duplicate call.

This holds true for all the recreate methods.

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java (hunk 4 above)
Comment (rish691) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java]: Done

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java [hunk 7]
```diff
@@ -87,22 +84,18 @@ private void buildParentChildRelationship(StructuredTrace trace) {
             .add(childEntity);
         childToParentEntities.computeIfAbsent(childEntity.getEntityId(), k -> new ArrayList<>())
             .add(parentEntity);
+        childrenEntityIds.add(childEntity.getEntityId());
```
Comment (avinashkolluru) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java:87]: Do we need this extra field. Can't we just use childToParentEntities.keySet() in `processEntities`?

Comment (avinashkolluru) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java:87]: Given that `childrenEntityIds` is not exposed in any form, i see no reason for just maintaining that extra set.

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java [hunk 8]
```diff
@@ -82,22 +80,19 @@ private void buildParentChildRelationship(StructuredTrace trace) {
         parentToChildrenEvents.computeIfAbsent(parentEvent.getEventId(), k -> new ArrayList<>())
             .add(childEvent);
         childToParentEvents.put(childEvent.getEventId(), parentEvent);
+        childrenEventIds.add(childEvent.getEventId());
```
Comment (avinashkolluru) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java:83]: same here too

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java (hunk 7 above)
Comment (rish691) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEntitiesGraph.java]: yeah not needed, missed it. Removed

**Review by avinashkolluru — APPROVED** [review]

**Pull request merged**